core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnEncoding.scala

```diff
@@ -0,0 +1,346 @@
+
+  def readBoolean(columnBytes: Array[Byte]): Boolean
+
+  def nextBoolean(columnBytes: Array[Byte]): Unit
+
+  def readByte(columnBytes: Array[Byte]): Byte
+
+  def nextByte(columnBytes: Array[Byte]): Unit
+
+  def readShort(columnBytes: Array[Byte]): Short
+
+  def nextShort(columnBytes: Array[Byte]): Unit
+
+  def readInt(columnBytes: Array[Byte]): Int
+
+  def nextInt(columnBytes: Array[Byte]): Unit
+
+  def readLong(columnBytes: Array[Byte]): Long
+
+  def nextLong(columnBytes: Array[Byte]): Unit
+
+  def readFloat(columnBytes: Array[Byte]): Float
+
+  def nextFloat(columnBytes: Array[Byte]): Unit
+
+  def readDouble(columnBytes: Array[Byte]): Double
+
+  def nextDouble(columnBytes: Array[Byte]): Unit
+
+  def readDecimal(columnBytes: Array[Byte], precision: Int,
+      scale: Int): Decimal
+
+  def nextDecimal(columnBytes: Array[Byte], precision: Int): Unit
+
+  def readUTF8String(columnBytes: Array[Byte]): UTF8String
+
+  def nextUTF8String(columnBytes: Array[Byte]): Unit
+
+  def readBinary(columnBytes: Array[Byte]): Array[Byte]
+
+  def nextBinary(columnBytes: Array[Byte]): Unit
+
+  def readInterval(columnBytes: Array[Byte]): CalendarInterval
+
+  def nextInterval(columnBytes: Array[Byte]): Unit
+
+  def readArray(columnBytes: Array[Byte]): UnsafeArrayData
+
+  def readMap(columnBytes: Array[Byte]): UnsafeMapData
+
+  def readStruct(columnBytes: Array[Byte], numFields: Int): UnsafeRow
+
+  def initializeEncoding(dataType: DataType, batchSize: Int): Array[Byte]
+
+  def writeBoolean(columnBytes: Array[Byte], value: Boolean): Unit
+
+  def writeByte(columnBytes: Array[Byte], value: Byte): Unit
+
+  def writeShort(columnBytes: Array[Byte], value: Short): Unit
+
+  def writeInt(columnBytes: Array[Byte], value: Int): Unit
+
+  def writeLong(columnBytes: Array[Byte], value: Long): Unit
+
+  def writeFloat(columnBytes: Array[Byte], value: Float): Unit
+
+  def writeDouble(columnBytes: Array[Byte], value: Double): Unit
+
+  def writeDecimal(columnBytes: Array[Byte], value: Decimal,
+      precision: Int): Unit
+
+  def writeUTF8String(columnBytes: Array[Byte], value: UTF8String): Unit
+
+  def writeBinary(columnBytes: Array[Byte], value: Array[Byte]): Unit
+
+  def writeInterval(columnBytes: Array[Byte], value: CalendarInterval): Unit
+
+  def writeArray(columnBytes: Array[Byte], value: UnsafeArrayData): Unit
+
+  def writeMap(columnBytes: Array[Byte], value: UnsafeMapData): Unit
+
+  def writeStruct(columnBytes: Array[Byte], value: UnsafeRow): Unit
+}
+
+object ColumnEncoding {
+
+  private[columnar] val baseOffset: Int =
+    Platform.BYTE_ARRAY_OFFSET + 4 /* for typeId */
+
+  private[columnar] val bitSetWords: Field = {
+    val f = classOf[BitSet].getDeclaredField("words")
+    f.setAccessible(true)
+    f
+  }
+
+  private[columnar] val BITS_PER_LONG = 64
+
+  val littleEndian: Boolean = ByteOrder.nativeOrder == ByteOrder.LITTLE_ENDIAN
+
+  val allEncodings: Array[(DataType, Boolean) => ColumnEncoding] = Array(
+    createPassThrough,
```

Comment:
Hmm. Haven't thought too much about it but since we are discussing, should we release Snappy with different versions of Spark in future. Otherwise, it would be very limiting. For e.g. when we release Snappy 1.0 with Spark 2.1, we should also release Snappy 1.0 with Spark 2.0. Something like what Spark does with hadoop versions.